File: #0279.perfect-squares.cpp
class Solution {
public:
    int numSquares(int n) {

        int i;
        int min;
        vector<int> result = {0};
        while(result.size() <= n){
            i = result.size();
            min = INT_MAX;
            for (int j = 1; j * j <= i; ++j) {
                min = std::min(min, result[i - j * j] + 1);
            }
            result.push_back(min);
        }
        return result[n];
    }
};

File: #0279.perfect-squares.js
/**
 * @param {number} n
 * @return {number}
 */
var numSquares = function(n) {
    const dp = Array(n + 1).fill(n);
    dp[0] = 0;
    for (let i = 0; i <= n; i++) {
        for (let j = 1; j <= n; j++) {
            let square = j * j;
            if (square + i > n) {
                break;
            }

            dp[i + square] = Math.min( 1 + dp[i], dp[i + square]);
        }
    }
    return dp[n];
};
